cmd/eventkitd-sink-bigquery/main.go

```diff
@@ -0,0 +1,207 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"net"
+	"os"
+	"os/signal"
+	"runtime"
+	"syscall"
+	"time"
+
+	"cloud.google.com/go/bigquery"
+	"github.com/jtolio/eventkit/pb"
+	"github.com/jtolio/eventkit/transport"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+type Application struct {
+	Name     string
+	Instance string
+	Version  string
+}
+
+type Record struct {
+	Application Application
+
+	StartTime  time.Time
+	ReceivedAt time.Time
+	SendOffset time.Duration
+
+	Source *net.UDPAddr
+
+	Tags []*pb.Tag
+}
+
+func (r *Record) Save() (map[string]bigquery.Value, string, error) {
+	fields := make(map[string]bigquery.Value)
+	fields["application_name"] = r.Application.Name
+	fields["application_instance"] = r.Application.Instance
+	fields["application_version"] = r.Application.Version
+	fields["start_time"] = r.StartTime
+	fields["received_at"] = r.ReceivedAt
+	fields["send_offset"] = r.SendOffset
+	fields["source_ip"] = r.Source.IP.String()
+
+	for _, tag := range r.Tags {
+		field := "tag_" + tag.Key
+
+		switch v := tag.Value.(type) {
+		case *pb.Tag_Bool:
+			fields[field] = v.Bool
+		case *pb.Tag_Bytes:
+			fields[field] = v.Bytes
+		case *pb.Tag_Double:
+			fields[field] = v.Double
+		case *pb.Tag_Duration:
+			fields[field] = v.Duration
+		case *pb.Tag_Int64:
+			fields[field] = v.Int64
+		case *pb.Tag_String_:
+			fields[field] = v.String_
+		}
+	}
+
+	return fields, "", nil
+}
+
+// BigQuerySink provides an abstraction for processing events in a transport agnostic way.
+type BigQuerySink struct {
+	dataset *bigquery.Dataset
+}
+
+// Receive is called when the server receive an event to process.
+func (b *BigQuerySink) Receive(ctx context.Context, packet *Packet) error {
+	records := make(map[string][]*Record)
+
+	for _, event := range packet.Packet.Events {
+		for _, scope := range event.Scope {
```

Comment:
a scope is a single entity. like the scope might be []string{"github.com", "jtolio", "eventkit"}, or it might be []string{"github.com/jtolio/eventkit", "debugmetrics"}, etc. a for loop is surprising here. you're duplicating the record for each scope element, but a scope is really more like a path. if we must force this into a string, i'd suggest picking an element delimiter and joining, though what i did here https://github.com/jtolio/eventkit/blob/main/eventkitd/path.go was escape and encode the path elements before joining

Reply:
that makes way more sense